internal/cluster/partition_service.go

```diff
@@ -0,0 +1,168 @@
+		partitionTable:       defaultPartitionTable(),
+		logger:               bundle.Logger,
+	}
+}
+
+func (s *PartitionService) Start() {
+	subscriptionID := event.MakeSubscriptionID(s.handlePartitionsUpdated)
+	s.eventDispatcher.Subscribe(EventPartitionsUpdated, subscriptionID, s.handlePartitionsUpdated)
+}
+
+func (s *PartitionService) Stop() {
+	subscriptionID := event.MakeSubscriptionID(s.handlePartitionsUpdated)
+	s.eventDispatcher.Unsubscribe(EventPartitionsUpdated, subscriptionID)
+}
+
+func (s *PartitionService) GetPartitionOwner(partitionId int32) internal.UUID {
+	return s.partitionTable.GetOwnerUUID(partitionId)
+}
+
+func (s *PartitionService) PartitionCount() int32 {
+	return atomic.LoadInt32(&s.partitionCount)
+}
+
+func (s *PartitionService) GetPartitionID(keyData pubserialization.Data) int32 {
+	if count := s.PartitionCount(); count == 0 {
+		// Partition count can not be zero for the sync mode.
+		// On the sync mode, we are waiting for the first connection to be established.
+		// We are initializing the partition count with the value coming from the server with authentication.
+		// This exception is used only for async mode client.
+		// TODO: panic
+		//core.NewHazelcastClientOfflineException()
+		return 0
+	} else {
+		return murmur.HashToIndex(keyData.PartitionHash(), count)
+	}
+}
+
+func (s *PartitionService) GetPartitionIDWithKey(key interface{}) (int32, error) {
+	if data, err := s.serializationService.ToData(key); err != nil {
+		return 0, err
+	} else {
+		return s.GetPartitionID(data), nil
+	}
+}
+
+func (s *PartitionService) handlePartitionsUpdated(event event.Event) {
+	if partitionsUpdatedEvent, ok := event.(*PartitionsUpdated); ok {
+		s.partitionTable.Update(partitionsUpdatedEvent.Partitions, partitionsUpdatedEvent.Version)
+		s.eventDispatcher.Publish(NewPartitionsLoaded())
+		s.logger.Debug(func() string { return "partitions loaded" })
+	}
+}
+
+func (s *PartitionService) checkAndSetPartitionCount(newPartitionCount int32) {
+	atomic.CompareAndSwapInt32(&s.partitionCount, 0, newPartitionCount)
+}
+
+type partitionTable struct {
+	partitionStateVersion int32
+	partitions            map[int32]internal.UUID
+	mu                    *sync.RWMutex
+}
+
+func (p *partitionTable) Update(pairs []proto.Pair, version int32) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	if version > p.partitionStateVersion {
+		newPartitions := map[int32]internal.UUID{}
+		for _, pair := range pairs {
+			uuids := pair.Key().([]internal.UUID)
+			ids := pair.Value().([]int32)
+			for _, uuid := range uuids {
+				for _, id := range ids {
+					newPartitions[id] = uuid
+				}
+			}
+		}
+		p.partitions = newPartitions
+		p.partitionStateVersion = version
+	}
+}
+
+func (p *partitionTable) GetOwnerUUID(partitionID int32) internal.UUID {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	if uuid, ok := p.partitions[partitionID]; ok {
+		return uuid
+	}
+	return nil
+}
+
+func (p *partitionTable) PartitionCount() int {
```

Comment:
This method looks unused. I think we can delete it. Partition count is already a different variable.

Reply:
This method is used in the GetPartitionID function. We cannot unexport PartitionCount, since there is also `partitionCount`. This code is package private, so we can keep it as is for now.